src/llamafactory/train/dpo/workflow.py

```diff
@@ -79,6 +86,14 @@ def run_dpo(
     # Training
     if training_args.do_train:
         train_result = trainer.train(resume_from_checkpoint=training_args.resume_from_checkpoint)
+        train_result.metrics["effective_tokens_per_sec"] = (
+            effective_token_num * train_result.metrics["epoch"] / train_result.metrics["train_runtime"]
+        )
+        if dist.is_initialized():
+            train_result.metrics["effective_tokens_per_sec"] = (
+                train_result.metrics["effective_tokens_per_sec"] / dist.get_world_size()
+            )
```

Comment:
Could we wrap these lines in a util function in llamafactory.extras.misc?

Reply:
done